noir-projects/aztec-nr/aztec/src/utils/bytes.nr

```diff
@@ -1,88 +1,307 @@
+    let mut fields = [0; (N + 30) / 31];
+
+    // There are N / 31 whole fields that can be populated.
+    for i in 0..N / 31 {
+        let mut field = 0;
+        for j in 0..31 {
+            let k = 31 - j - 1; // iterate in reverse.
+            // Shift the existing value left by 8 bits and add the new byte
+            field = field * 256 + bytes[i * 31 + k] as Field;
         }
+        fields[i] = field;
+    }
 
-        dst[dst_index] = field_value;
+    // Note: if 31 divides N, then this loop does not execute.
+    // ceil(N/31) - floor(N/31) = 1, unless 31 divides N, in which case it's 0.
+    for _ in 0..((N + 30) / 31) - (N / 31) {
+        let mut final_field = 0;
+        let final_field_idx = fields.len() - 1;
+        let final_offset = final_field_idx * 31;
+        // N - ((N / 31) * 31) = N - floor(N/31) * 31 = the number of bytes to go in the final field.
+        for j in 0..N - ((N / 31) * 31) {
+            let k = (N - ((N / 31) * 31)) - j - 1; // iterate in reverse.
+            // Shift the existing value left by 8 bits and add the new byte
+            final_field = final_field * 256 + bytes[final_offset + k] as Field;
+        }
+        fields[final_field_idx] = final_field;
     }
 
-    dst
+    fields
 }
 
 // Converts an input array of fields into bytes. Each field of input has to contain only 31 bytes.
-// TODO(#8618): Optimize for public use.
-pub fn fields_to_bytes<let N: u32, let M: u32>(input: [Field; M]) -> [u8; N] {
-    let mut dst = [0; N];
+// It appears that the original author of this function wants to give the caller some control
+// over the number of bytes that are returned: the final field can be mapped to fewer than 31 bytes, it seems,
+// based on the caller's choice of M.
+// Given the input is N fields, and each field (except apparently the last) is mapped to 31 bytes,
+// we definitely need M > (N - 1) * 31 to cover those first (N - 1) fields. Hence the static assertion.
+// TODO: this is the same as the "le" version of this function, save for the `field.to_be_bytes()` calls. I tried passing the conversion function in as a parameter, to reduce code duplication, but couldn't get it to work.
+pub fn fields_to_be_bytes_31<let N: u32, let M: u32>(fields: [Field; N]) -> [u8; M] {
+    let mut bytes = [0; M];
+
+    std::static_assert(M > (N - 1) * 31, "Not enough output bytes specified");
+    // We make this assertion to keep the logic simpler:
+    std::static_assert(N * 31 >= M, "Too many output bytes specified.");
+
+    for i in 0..N - 1 {
+        let field = fields[i];
+
+        // We expect that the field contains at most 31 bytes of information.
+        field.assert_max_bit_size::<248>();
+
+        // Now we can safely convert the field to 31 bytes.
+        let field_as_bytes: [u8; 31] = field.to_be_bytes();
+
+        for j in 0..31 {
+            bytes[i * 31 + j] = field_as_bytes[j];
+        }
+    }
+
+    // The final field will be
```

Comment:
Missing comment?

Reply:
The final field _will_ be, Nico. The final field will be.

(Deleted. Not sure what I was going to say).